File: cracking_interview/ch1/q2_check_permutation.exs
ExUnit.start()

defmodule Ch1 do
  @moduledoc """
  Given two string, check if one is permutation of the other
  """

  @doc "Using sorting"
  def permutation?(one, other) when length(one) != length(other), do: False
  def permutation?(one, other) do
    Enum.sort(String.graphemes one) == Enum.sort(String.graphemes other)
  end

  @doc "counting codepoints"
  def permutation2?(one, other) when length(one) != length(other), do: False
  def permutation2?(one, other), do: count(one) == count(other)
  
  defp count(string, counter \\ %{})
  defp count("", counter), do: counter
  defp count(<<point::utf8, rest::bitstring>>, counter) do
    counter = elem(Map.get_and_update(counter, point, fn
      nil -> {nil, 1}
      value -> {value, value + 1}
    end), 1)
    count(rest, counter)
  end
end

defmodule Ch1Test do
  use ExUnit.Case, async: true

  test "Is permutation" do
    assert Ch1.permutation?("abcd", "dcba")
  end

  test "Is not permutation length" do
    assert !Ch1.permutation?("abcd", "dcb")
  end

  test "Is not permutation" do
    assert !Ch1.permutation?("abcd", "dcbe")
  end

  test "Is permutation2" do
    assert Ch1.permutation2?("abcd", "dcba")
  end

  test "Is not permutation2 length" do
    assert !Ch1.permutation2?("abcd", "dcb")
  end

  test "Is not permutation2" do
    assert !Ch1.permutation2?("abcd", "dcbe")
  end
end

File: cracking_interview/ch1/q2_check_permutation.py
"""
Given two string, check if one is permutation of the other
"""

def check(one, other):
    """
    Using sorting ((a+b)log(a*b))

    >>> check("abcda", "aadcb")
    True
    >>> check("abc", "bcd")
    False
    >>> check("a", "bc")
    False
    """
    if len(one) == len(other):
        return sorted(one) == sorted(other)
    return False

def count(string, max_char):
    """
    Count occurrences

    >>> count("\x02\x03\x02", 5)
    [0, 0, 2, 1, 0]
    """
    state = [0] * max_char
    for char in string:
        state[ord(char)] += 1
    return state


def check2(one, other):
    """
    Counting occurrences. 3(a+b) + c -> O(N)

    >>> check2("abcda", "aadcb")
    True
    >>> check2("abc", "abd")
    False
    >>> check2("abc", "bc")
    False
    """
    if len(one) != len(other):
        return False
    size = max(ord(x) for x in one + other) + 1

    res_one = count(one, size)
    res_other = count(other, size)

    for i in range(size):
        if res_one[i] != res_other[i]:
            return False
    return True
